src/powerset.rs

```diff
@@ -0,0 +1,85 @@
+use std::fmt;
+use alloc::vec::Vec;
+use std::usize;
+
+use super::combinations::Combinations;
+use super::size_hint;
+
+/// An iterator over the powerset of all elements from its source iterator.
+///
+/// See [`.powerset()`](../trait.Itertools.html#method.powerset) for more
+/// information.
+#[must_use = "iterator adaptors are lazy and do nothing unless consumed"]
+pub struct Powerset<I: Iterator> {
+    combs: Combinations<I>,
+    pos: usize,
+}
+
+impl<I> Clone for Powerset<I>
+    where I: Clone + Iterator,
+          I::Item: Clone,
+{
+    clone_fields!(combs, pos);
+}
+
+impl<I> fmt::Debug for Powerset<I>
+    where I: Iterator + fmt::Debug,
+          I::Item: fmt::Debug,
+{
+    debug_fmt_fields!(Powerset, combs, pos);
+}
+
+/// Create a new `Powerset` from a cloneable iterator.
+pub fn powerset<I>(src: I) -> Powerset<I>
+    where I: Iterator,
+          I::Item: Clone,
+{
+    Powerset {
+        combs: Combinations::new(src, 0),
+        pos: 0,
+    }
+}
+
+impl<I> Iterator for Powerset<I>
+    where
+        I: Iterator,
+        I::Item: Clone,
+{
+    type Item = Vec<I::Item>;
+
+    fn next(&mut self) -> Option<Vec<I::Item>> {
+        if let Some(elt) = self.combs.next() {
+            self.pos = self.pos.saturating_add(1);
+            Some(elt)
+        } else {
+            if self.combs.k() < self.combs.n() || self.combs.k() == 0 {
+                self.combs.reset(self.combs.k() + 1);
+
+                let result = self.combs.next();
+                if result.is_some() {
+                    self.pos = self.pos.saturating_add(1);
+                }
+
+                result
+            } else {
+                None
+            }
```

Comment:
(Nitpick) Should we get rid of this level of indentation by using `else if self.combs.k()...` in line 54?

Reply:
Good idea!